### PR TITLE
UPSTREAM: <carry>: openshift: Add support for Custom data, read from secrets for Ignition

### PR DESCRIPTION
**Pull request by awesomenix**
…Secrets for Ignition

**What this PR does / why we need it**:

- Adds support for Customdata in vm
- Gets customdata from secret if provided
- Skip calling vm extension if custom data is provided

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Bootstrap using Custom Data specified in secrets
```

### CODE REVIEW

File: pkg/apis/azureprovider/v1alpha1/types.go [hunk 1]
```diff
@@ -369,10 +369,11 @@ type VM struct {
...
 type Image struct {
-	Publisher string `json:"publisher"`
-	Offer     string `json:"offer"`
-	SKU       string `json:"sku"`
-	Version   string `json:"version"`
+	Publisher  string `json:"publisher"`
+	Offer      string `json:"offer"`
+	SKU        string `json:"sku"`
+	Version    string `json:"version"`
+	ResourceID string `json:"resourceID"`
```
Comment (enxebre) [inline pkg/apis/azureprovider/v1alpha1/types.go]: how is this related to this PR/commit?

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 2]
```diff
@@ -502,5 +517,29 @@ func (s *Reconciler) createVirtualMachine(ctx context.Context, nicName string) e
 		}
 	}
 
-	return err
+	return nil
+}
+
+func (s *Reconciler) getCustomScriptData() (string, error) {
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: nit: getCustomUserData?

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 3]
```diff
@@ -502,5 +517,29 @@ func (s *Reconciler) createVirtualMachine(ctx context.Context, nicName string) e
...
+	scriptData := ""
+	if s.scope.MachineConfig.UserDataSecret != nil {
+		var userDataSecret apicorev1.Secret
+		cfg := controllerconfig.GetConfigOrDie()
+		cs, err := client.New(cfg, client.Options{})
+		if err != nil {
+			return "", errors.Wrapf(err, "error creating new client for secret %s in namespace %s", s.scope.MachineConfig.UserDataSecret.Name, s.scope.Namespace())
+		}
+
+		err = cs.Get(context.Background(), client.ObjectKey{Namespace: s.scope.Namespace(), Name: s.scope.MachineConfig.UserDataSecret.Name}, &userDataSecret)
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: nit: we could inline here `if err != nil` here

**Comment by enxebre** [comment]
looks great
PTAL @frobware @ingvagabund

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 4]
```diff
@@ -94,19 +91,26 @@ func (s *Reconciler) Create(ctx context.Context) error {
...
+		if err != nil {
+			return errors.Wrapf(err, "failed to get vm startup script")
+		}
+
+		vmExtSpec := &virtualmachineextensions.Spec{
+			Name:       "startupScript",
+			VMName:     s.scope.Machine.Name,
+			ScriptData: base64.StdEncoding.EncodeToString([]byte(scriptData)),
+		}
+		err = s.virtualMachinesExtSvc.CreateOrUpdate(ctx, vmExtSpec)
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: Simplify to 

```go
if err := s.virtualMachinesExtSvc.CreateOrUpdate(ctx, vmExtSpec); err != nil {
...
```

**Comment by frobware** [comment]
Could we change the title and add back "Secrets for Ignition" that github unhelpfully removed.

File: pkg/apis/azureprovider/v1alpha1/types.go (hunk 1 above)
Comment (ingvagabund) [reply pkg/apis/azureprovider/v1alpha1/types.go]: +1, it does not seem the new field has anything to do with allowing to inject user defined data. Can you open a separate PR for it? Resp. does it need to be implemented before the user data can be constructed?

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 5]
```diff
@@ -78,13 +80,8 @@ func (s *Reconciler) Create(ctx context.Context) error {
...
 
-	bootstrapToken, err := s.checkControlPlaneMachines()
-	if err != nil {
-		return errors.Wrap(err, "failed to check control plane machines in cluster")
-	}
-
 	nicName := fmt.Sprintf("%s-nic", s.scope.Machine.Name)
-	err = s.createNetworkInterface(ctx, nicName)
+	err := s.createNetworkInterface(ctx, nicName)
 	if err != nil {
```
Comment (ingvagabund) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: Just a recommendation:
```go
if err := s.createNetworkInterface(ctx, nicName); err != nil {...}
```

Though, it's still preferable so one does not test `err` variable of a different expressions. The smaller scope of a variable the better.

File: pkg/apis/azureprovider/v1alpha1/types.go (hunk 1 above)
Comment (awesomenix) [reply pkg/apis/azureprovider/v1alpha1/types.go]: Agreed doesnt have to be part of the PR, fixing

File: pkg/cloud/azure/actuators/machine/reconciler.go (hunk 4 above)
Comment (awesomenix) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: unfortunately linter complains of creating a overshadowing variable for if err := ; err != nil pattern hence i went this route :( I dont want to make mass changes since we want to merge upstream

File: pkg/cloud/azure/actuators/machine/actuator_test.go [hunk 6]
```diff
@@ -470,6 +472,44 @@ func TestAvailabilityZones(t *testing.T) {
...
+		Data: map[string][]byte{
+			"userData": []byte("test-userdata"),
+		},
+	}
+	fakeScope.CoreClient = controllerfake.NewFakeClient(userDataSecret)
+	fakeScope.MachineConfig.UserDataSecret = &corev1.SecretReference{Name: "testCustomUserData"}
+	fakeReconciler := newFakeReconcilerWithScope(t, fakeScope)
+	fakeReconciler.virtualMachinesSvc = &FakeVMCheckZonesService{}
+	if err := fakeReconciler.Create(context.Background()); err != nil {
+		t.Errorf("expected create to succeed %v", err)
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator_test.go]: Is this actually `Fatal`? Can the remainder of the tests run gracefully if this fails?

File: pkg/cloud/azure/actuators/machine/actuator_test.go [hunk 7]
```diff
@@ -470,6 +472,44 @@ func TestAvailabilityZones(t *testing.T) {
...
+	}
+
+	userData, err := fakeReconciler.getCustomUserData()
+	if err != nil {
+		t.Errorf("expected get custom data to succeed %v", err)
+	}
+
+	if userData != "test-userdata" {
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/actuator_test.go]: This only tests the positive - can we create (perhaps table-driven tests) for success and expected failure?

File: pkg/cloud/azure/actuators/machine/actuator_test.go (hunk 6 above)
Comment (awesomenix) [reply pkg/cloud/azure/actuators/machine/actuator_test.go]: Yes it does run gracefully, but it would fail (depending on the tests), in this case it would report list of failures instead of failing at the first test itself.

File: pkg/cloud/azure/actuators/machine/actuator_test.go (hunk 7 above)
Comment (awesomenix) [reply pkg/cloud/azure/actuators/machine/actuator_test.go]: The tests do have negative tests right below this one. I did try to split out positive and negative, but didnt feel the need to 2 additional statements, i didnt cover one additional case of missing "userdata" field in secret, let me do that right away

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 8]
```diff
@@ -474,6 +477,15 @@ func (s *Reconciler) createVirtualMachine(ctx context.Context, nicName string) e
 			Zone:       vmZone,
 		}
 
+		userData, userDataErr := s.getCustomUserData()
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: >// CustomData - Specifies a base-64 encoded string of custom data. The base-64 encoded string is decoded to a binary array that is saved as a file on the Virtual Machine. The maximum length of the binary array is 65535 bytes. <br><br> For using cloud-init for your VM, see [Using cloud-init to customize a Linux VM during creation](https://docs.microsoft.com/azure/virtual-machines/virtual-machines-linux-using-cloud-init?toc=%2fazure%2fvirtual-machines%2flinux%2ftoc.json)

This would need to go through `base64.StdEncoding.EncodeToString() to be consumable by `OSProfile.CustomData`

Comment (awesomenix) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: Nice catch, i was referring to the parent PR and i missed porting it over, fixing it now and adding test to catch this

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 9]
```diff
@@ -502,5 +514,23 @@ func (s *Reconciler) createVirtualMachine(ctx context.Context, nicName string) e
...
+	if s.scope.MachineConfig.UserDataSecret != nil {
+		var userDataSecret apicorev1.Secret
+
+		if err := s.scope.CoreClient.Get(context.Background(), client.ObjectKey{Namespace: s.scope.Namespace(), Name: s.scope.MachineConfig.UserDataSecret.Name}, &userDataSecret); err != nil {
+			return "", errors.Wrapf(err, "error getting user data secret %s in namespace %s", s.scope.MachineConfig.UserDataSecret.Name, s.scope.Namespace())
+		}
+		if data, exists := userDataSecret.Data["userData"]; exists {
+			userData = string(data)
+		} else {
+			klog.Warningf("Secret %v/%v does not have userData field set. Thus, no user data applied when creating an instance.", s.scope.Namespace(), s.scope.MachineConfig.UserDataSecret.Name)
```
Comment (enxebre) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: I know this is coming from aws but I think we should probably just fail here

**Comment by enxebre** [comment]
@awesomenix could you put the `vendor/Gopkg` changes in their own commit? other than that and https://github.com/openshift/cluster-api-provider-azure/pull/13/files#r273853597 lgtm

File: pkg/cloud/azure/actuators/machine/reconciler.go (hunk 9 above)
Comment (awesomenix) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: I thought the reason why it wouldnt want to fail is to support not running bootstrap ignition.  But i cant think of a case of that now. Returning error and fixing tests

**Comment by enxebre** [comment]
@frobware @ingvagabund @bison PTAL
/hold cancel

File: pkg/cloud/azure/actuators/machine/reconciler.go [hunk 10]
```diff
@@ -502,5 +514,23 @@ func (s *Reconciler) createVirtualMachine(ctx context.Context, nicName string) e
 		}
 	}
 
-	return err
+	return nil
+}
+
+func (s *Reconciler) getCustomUserData() (string, error) {
+	userData := ""
+	if s.scope.MachineConfig.UserDataSecret != nil {
```
Comment (frobware) [inline pkg/cloud/azure/actuators/machine/reconciler.go]: This is a nit for me in this PR now... but if you invert the test, the indentation reduces and the other `if...else` could be simpler too.

Comment (awesomenix) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: I couldnt figure out a way to fixup previous commit, created a new one for simplicity, hope thats ok

**Comment by frobware** [comment]
/lgtm

Thanks for the update in getCustomUserData().

File: pkg/cloud/azure/actuators/machine/reconciler.go (hunk 10 above)
Comment (enxebre) [reply pkg/cloud/azure/actuators/machine/reconciler.go]: FYI you can rewrite and edit your e.g last 2 commits with `git rebase -i HEAD~2`
Additionally you might also find useful `git stash save/pop`
